nucypher/utilities/latency.py

```diff
@@ -0,0 +1,84 @@
+import time
+from collections import defaultdict
+from threading import Lock
+from typing import List
+
+from eth_typing import ChecksumAddress
+
+
+class NodeLatencyStatsCollector:
+    """
+    Thread-safe utility that tracks latency statistics related to P2P connections with other nodes.
+    """
+
+    CURRENT_AVERAGE = "current_avg"
+    COUNT = "count"
+    MAX_LATENCY = float(2**16)  # just need a large number for sorting
+
+    class NodeLatencyContextManager:
+        def __init__(
+            self,
+            stats_collector: "NodeLatencyStatsCollector",
+            staker_address: ChecksumAddress,
+        ):
+            self._stats_collector = stats_collector
+            self.staker_address = staker_address
+
+        def __enter__(self):
+            self.start_time = time.perf_counter()
+            return self
```

Comment:
Is it necessary to use to lock in here (and `__exit__`) too? Or in other words, is it possible for `self.start_time` to become a shared resource?

Reply:
> is it possible for self.start_time to become a shared resource?

It's not possible in the way we currently use it. 

However, it's generally possible, but seems more of a misuse of the context manager instance.

i.e. the expected usage is generally
```python
with node_latency_collector.get_latency_tracker(<>):
    ...
```
i.e. a fresh instance of the context manager is used when needed. Fresh instances prevent the the shared resource since they are self-contained.

The scenarios I can think of for `start_time` to become a shared resource would be:
1. nesting of the context manager. e.g.
```python
# get and store instance
tracker = node_latency_collector.get_latency_tracker(<>)

with tracker:
     ....
     with tracker:
         ...
```

so the nesting causes the original start time from the outside block to be changed when the internal block is hit. This is particularly ugly and shouldn't be done.

OR 

2. the use of the same instance in multiple threads:
```python
# get and store instance
tracker = node_latency_collector.get_latency_tracker(<>)

# use tracker instance within thread_1
# use same tracker instance within thread_2
...
```
So, you would need to have a common context manager instance and then use that instance concurrently in separate threads.

Reply:
Right, I think I was confused by the use of locks to handle stats, but then I realized this is due to the stats collector being the potentially shared resource.